presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java

```diff
@@ -189,6 +210,49 @@ private void enforceTimeLimits()
         }
     }
 
+    /**
+     *  When cluster reaches max tasks limit and also a single query
+     *  excceds a threshold,  kill this query
+     */
+    private void enforceTaskLimits()
+    {
+        int totalRunningTaskCount = 0;
+        int highestRunningTaskCount = 0;
+        Optional<T> highestRunningTaskQuery = Optional.empty();
+        for (T query : queries.values()) {
+            if (query.isDone()) {
+                continue;
+            }
+            int runningTaskCount = query.getRunningTask();
+            totalRunningTaskCount += runningTaskCount;
+            if (runningTaskCount > highestRunningTaskCount) {
+                highestRunningTaskCount = runningTaskCount;
+                highestRunningTaskQuery = Optional.of(query);
+            }
+        }
+
+        runningTaskCount.set(totalRunningTaskCount);
+
+        if (totalRunningTaskCount > maxTotalRunningTaskCount &&
+                highestRunningTaskCount > maxQueryRunningTaskCount &&
+                highestRunningTaskQuery.isPresent()) {
+            highestRunningTaskQuery.get().fail(new PrestoException(QUERY_HAS_TOO_MANY_STAGES, format(
+                    "Cluster is overloaded with too many tasks (%s), query with highest running tasks is killed (%s). " + TOO_MANY_STAGES_MESSAGE,
```

Comment:
This format string uses the wrong modifiers and mixes formatting with concatenation. I assume this doesn't have unit test coverage because if it did it would fail trying to format this message?

Reply:
The old way works, but good point to keep it consistent

Reply:
Oh interesting. Since numbers are boxed they have a toString that does the right thing. TIL.